### PR TITLE
fix(ci): epoll on pidfd to wait for Firecracker exit

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Substitute this mechanism with calling epoll() on the pidfd of the process instead. This should deterministically block until the process exits. If there's something else wrong, we will hit the pytest timeout.

## Reason

Currently, we use psutil.pid_exists in a loop with a timeout of 10 seconds. This is racy and indeed some times we hit it in our CI.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this
  PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet
  [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tests/framework/utils.py [hunk 1]
```diff
@@ -450,16 +451,38 @@ def run_guest_cmd(ssh_connection, cmd, expected, use_json=False):
...
     is still alive after retrying several times.
     """
-    if psutil.pid_exists(p_pid):
-        raise Exception(f"[{p_pid}] process is still alive")
+    pidfd = get_process_pidfd(p_pid)
+
+    # If pidfd is None the process has already terminated
+    if pidfd is not None:
+        epoll = select.epoll()
+        epoll.register(pidfd, select.EPOLLIN)
```
Comment (kalyazin) [inline tests/framework/utils.py:483]: if the process goes between `get_process_pidfd` and `register`, will we still receive the `EPOLLIN`?

Comment (bchalios) [reply tests/framework/utils.py:483]: I _think_ so. IIUC, if `pidfd_open` succeeds it opens a pidfd file and installs the file descriptor in the caller's file descriptor table. This means that we can poll on it until we close it.

Also, when we `poll` on it, we end up doing this: https://elixir.bootlin.com/linux/v6.11.3/source/fs/pidfs.c#L109, which checks for the `struct task` that corresponds to the `pid` that corresponds to the `pidfd`. If this is NULL, the system call returns `EPOLLIN` directly.

Comment (bchalios) [reply tests/framework/utils.py]: I've changed this to use `select.select` directly

**Review by ShadowCurse — DISMISSED** [review]

**Comment by bchalios** [comment]
@ShadowCurse I had to revert back to using `epoll()`. `select()` hits the `FD_SETSIZE` issue (1024 descriptors only).

**Comment by ShadowCurse** [comment]
> @ShadowCurse I had to revert back to using `epoll()`. `select()` hits the `FD_SETSIZE` issue (1024 descriptors only).

How can it hit the limit if you are waiting for 1 fd?

**Comment by bchalios** [comment]
> How can it hit the limit if you are waiting for 1 fd?

It's not about the numbers of PIDs you're waiting on. It's about the maximum PID value it can handle. Reading from `man select(2)`:

>        select() can monitor only file descriptors numbers that are less than FD_SETSIZE; poll(2) and epoll(7) do not have this limitation.  See BUGS.

**Comment by roypat** [comment]
Since we only use `utils.wait_for_termination` in two locations, and both of those are in `Microvm`, can we move this function into `Microvm` and create the `pidfd` already when we spawn the Firecracker process? Otherwise, how does this behave if the Firecracker process dies unexpectedly and the pid gets freed significantly before we get to the `wait_for_termination` call? :thinking:

**Comment by bchalios** [comment]
> Otherwise, how does this behave if the Firecracker process dies unexpectedly and the pid gets freed significantly before we get to the wait_for_termination call? 🤔

If the process with pid `PID` is already dead when creating the `pidfd`, then the system call will raise a `ProcessLookupError` exception and `get_process_pidfd` will return `None`, so `wait_for_termination` will return immediately.

> can we move this function into Microvm and create the pidfd already when we spawn the Firecracker process?

I thought of doing that, but I think it makes the error handling much more complicated.

**Comment by ShadowCurse** [comment]
Ok, but to avoid issue with 2 calls: `register` and `epoll`, maybe we should use simple `poll`?

**Comment by bchalios** [comment]
`select.poll()` returns you a "polling" object to which you register file descriptors.

**Review by roypat — APPROVED** [review]

**Review by ShadowCurse — APPROVED** [review]

**Review by kalyazin — APPROVED** [review]

**Pull request merged**